api/src/trace.rs

```diff
@@ -63,3 +64,72 @@ pub enum ExecutionEvent {
         msg: String,
     },
 }
+
+impl From<ExecutionTrace> for InvocationTrace {
+    fn from(e_trace: ExecutionTrace) -> InvocationTrace {
+        let mut invocation_stack: Vec<InvocationTrace> = Vec::new();
+
+        for event in e_trace.clone().events.into_iter() {
+            match event {
+                ExecutionEvent::GasCharge { .. } | ExecutionEvent::Log { .. } => {}
+                ExecutionEvent::Call { from, to, method, params, value } => {
+                    invocation_stack.push(InvocationTrace {
+                        from: Address::new_id(from),
+                        to,
+                        method,
+                        params,
+                        value,
+                        code: ExitCode::OK, // Placeholder, will be updated during return
+                        ret: None,          // Placeholder, will be updated during return
+                        subinvocations: Vec::new(),
+                    });
+                }
+                ExecutionEvent::CallReturn { return_value, exit_code } => {
+                    let mut current_invocation = invocation_stack
+                        .pop()
+                        .unwrap_or_else(|| panic!("Unmatched call return: {:?}", e_trace));
+
+                    current_invocation.code = exit_code;
+                    current_invocation.ret = return_value;
+
+                    if let Some(parent_invocation) = invocation_stack.last_mut() {
+                        parent_invocation.subinvocations.push(current_invocation);
+                    } else {
+                        // At root invocation
+                        return current_invocation;
+                    }
```

Comment:
Please remove the early returns here and below, and instead assign to a root call. Then check the stack is empty after the loop finishes.